src/metabase/upload.clj

```diff
@@ -678,7 +676,54 @@
 ;;; |  hydrate based_on_upload for FE
 ;;; +--------------------------------
 
-(mi/define-simple-hydration-method based-on-upload
+(defn uploadable-table-ids
+  "Returns the subset of table ids where the user can upload to the table."
+  [table-ids]
+  (if (empty? table-ids)
+    #{}
+    (let [tables (t2/hydrate (t2/select :model/Table :id [:in table-ids]) :db)]
+      (set (keep (fn [t]
+                   (when (can-upload-to-table? (:db t) t)
+                     (:id t)))
+                 tables)))))
+
+(defn- no-joins?
+  "Returns true if `query` has no joins in it, otherwise false."
+  [query]
+  (let [all-joins (mapcat (fn [stage]
+                            (lib/joins query stage))
+                          (range (lib/stage-count query)))]
+    (empty? all-joins)))
+
+(mu/defn model-hydrate-based-on-upload
+  "Batch hydrates `:based_on_upload` for each item of `models`. Assumes each item of `model` represents a model."
+  [models :- [:sequential [:map
+                           ;; query_type and dataset_query can be null in tests, so we make them nullable here.
+                           ;; they should never be null in production
+                           [:dataset_query [:maybe ms/Map]]
+                           [:query_type    [:maybe [:or :string :keyword]]]
+                           [:table_id      [:maybe ms/PositiveInt]]
+                           ;; is_upload can be provided for an optional optimization
+                           [:is_upload {:optional true} [:maybe :boolean]]]]]
+  (let [table-ids            (->> models
+                                  ;; as an optimization when listing collection items (GET /api/collection/items),
+                                  ;; we might already know that the table is not an upload if is_upload=false. We
+                                  ;; can skip making more queries if so
+                                  (remove #(false? (:is_upload %)))
+                                  (keep :table_id)
+                                  set)
+        uploadable-table-ids (set (uploadable-table-ids table-ids))
+        based-on-upload      (fn [model]
+                               (when-let [dataset_query (:dataset_query model)] ; dataset_query is sometimes null in tests
+                                 (let [query (lib/->pMBQL dataset_query)]
+                                   (when (and (= (name (:query_type model)) "query")
```

Comment:
Should this not be "failing closed" for nil query_type? `(= (name (:query_type model "query")) "query")`